### PR TITLE
Fix a nasty bug in the bytecode VM initialisation

### LINKED ISSUE #13512: GC crashes under bytecode
The [Gc torture test](https://github.com/ocaml-multicore/multicoretests/pull/469) has come to a point, where it can trigger a Gc crashing bug with the bytecode interpreter. This can trigger on both on `5.2.0`, on the `5.3` branch, and on `trunk`. I have not tried on 5.0.0 or 5.1.1 yet.

The test itself triggers list/string/array/bigarray allocations in both the parent domain and in two child domains and checks consistency of the results from `Gc.quick_stat`, etc. In between test iterations, we explicitly call `Gc.major` in single-domain mode in an attempt to reset the heap to a reasonable state. The test also uses a single call to a stub at start-up to support testing with `OCAMLRUNPARAM="s=4096"` set, but that can removed if unused.

This issue may or may not be the same bug observed in #13402.
As this test doesn't involve the `Obj`-using `Domain.DLS` I've gone for reporting it separately.

On GitHub actions CI I've observed these crashes with Ubuntu and glibc. For an example 5.3 crash see here:
https://github.com/ocaml-multicore/multicoretests/actions/runs/11074833760/job/30774502500?pr=469
Locally it seems easier to reproduce with musl C though.

To reproduce:
```
$ sudo apt-get install musl-tools

$ opam switch create . --empty
$ opam install . --inplace-build ocaml-option-musl ocaml-option-bytecode-only
$ opam install qcheck-core

$ git clone -b gc-test-musl-bytecode-repro https://github.com/ocaml-multicore/multicoretests.git
$ cd multicoretests
$ OCAMLRUNPARAM="o=20" dune exec src/gc/stm_tests_impl.exe --profile=debug-runtime -- -v
```

Above I suggest running with a reduced `space_overhead` and the debug runtime, but the bug can also trigger without these.

Here's a the result of a fresh run on my Linux box:
```
$ OCAMLRUNPARAM="o=20" dune exec src/gc/stm_tests_impl.exe --profile=debug-runtime -- -v
### OCaml runtime: debug mode ###   
### set OCAMLRUNPARAM=v=0 to silence this message
Page size: 4096
random seed: 57333975
generated error fail pass / total     time test name
[ ]  874    0    0  874 / 1000   500.8s STM implicit Gc test parallelBus error (core dumped)
```

with the following, not-so-descriptive stack trace:
```
(gdb) thread apply all bt

Thread 3 (LWP 1044096):
#0  0x00007fa68ff14b8c in memset () from /lib/ld-musl-x86_64.so.1
Backtrace stopped: Cannot access memory at address 0x7fa67cec89a8

Thread 2 (LWP 956593):
#0  0x00007fa68ff14bf8 in memset () from /lib/ld-musl-x86_64.so.1
#1  0x00007fa68ff05d35 in ?? () from /lib/ld-musl-x86_64.so.1
#2  0x00007fa67fc52b74 in ?? ()
#3  0x0000000000000000 in ?? ()

Thread 1 (LWP 956567):
#0  caml_interprete (prog=<optimized out>, prog_size=<optimized out>) at runtime/interp.c:829
#1  0x000060172030b6fd in caml_startup_code_exn (pooling=0, argv=0x7ffff57a4598, section_table_size=3833, section_table=0x60172032b020 <caml_sections> "\204\225\246\276", data_size=22927, data=0x60172032bf20 <caml_data> "\204\225\246\276", code_size=539264, code=0x6017203318c0 <caml_code>) at runtime/startup_byt.c:659
#2  caml_startup_code_exn (code=0x6017203318c0 <caml_code>, code_size=539264, data=0x60172032bf20 <caml_data> "\204\225\246\276", data_size=22927, section_table=0x60172032b020 <caml_sections> "\204\225\246\276", section_table_size=3833, pooling=0, argv=0x7ffff57a4598) at runtime/startup_byt.c:592
#3  0x000060172030b742 in caml_startup_code (code=code@entry=0x6017203318c0 <caml_code>, code_size=code_size@entry=539264, data=data@entry=0x60172032bf20 <caml_data> "\204\225\246\276", data_size=data_size@entry=22927, section_table=section_table@entry=0x60172032b020 <caml_sections> "\204\225\246\276", section_table_size=section_table_size@entry=3833, pooling=0, argv=0x7ffff57a4598) at runtime/startup_byt.c:673
#4  0x00006017202d5052 in main (argc=<optimized out>, argv=<optimized out>) at camlprim.c:26518
```

### PR DESCRIPTION
**Pull request by stedolan**
## The patch

The bytecode VM uses a fixed 7-instruction sequence of bytecode to set up a call from C into OCaml. This bytecode sequence requires initialisation, to register the code fragment (and thread the bytecode, if not a debug build).

Currently, this code sequence is defined per-thread and initialised on first use. This patch changes it to be defined globally and initialised at startup. (The main thread's copy was already initialised at startup through `caml_init_callbacks`. The change here is to make that the _only_ copy).

Using a code sequence per thread instead of one global one is somewhat less efficient but not by itself a bug. The problem is that the code fragments leaked, as they were not unregistered at thread termination. This causes #13512.

## The bug

With the old code, the following sequence of unfortunate events can occur, if you are very, very unlucky (or you are @jmid with multicoretests, manufacturing bad luck on an industrial scale):

  1. The program uses `Domain.spawn f`, which creates a new domain and uses `caml_callback_res` to invoke `f`.
  2. When `caml_callback_res` is called, the callback initialisation logic runs, creating a 28-byte code fragment. This fragment covers the 7-instruction callback sequence stored in thread-local storage.
  3. Later, the domain terminates. The C library / OS frees resources used by its threads, including the memory for thread-local storage. The 28-byte code fragment remains registered.
  4. Later, another domain does a minor GC, and runs out of space on the major heap to promote live blocks to. It requests more memory from the OS, which happens to return a 32K pool that includes the 28 bytes that are still registered as a code fragment. A particularly unlucky value is promoted into those 28 bytes.
  5. This unlucky value remains live, but happens not to be pointed to by any other heap value: all references to it are from local variables (so, on the bytecode stack).
  6. At the start of the next major GC cycle, the runtime scans the stack. The bytecode stack contains an arbitrary mix of values and code pointers (return addresses & exception handlers), so the GC skips over any pointers it finds that lie inside code fragments. So, it does not see that the unlucky value is live.
  7. The unlucky value gets collected, despite being in use, causing chaos when it's next used.

In other words: dangling code fragments that are not unregistered are not just a resource leak but a soundness issue. They leave blindspots in the address space that the GC cannot perceive and cause weird bugs later on.

### CODE REVIEW

**Comment by ghost** [comment]
This diff appears to solve indeed the bytecode issues (#13402 and probably others) for me.

However, are you sure it is safe to use a single copy for the whole program, as two elements of `callback_code` get modified everytime `caml_callbackN_exn` is invoked? In other words, are you sure there is no risk of two domains attempting to modify and use `callback_code` at the same time?

**Comment by stedolan** [comment]
Urgh, well spotted, I'd forgotten it actually mutates it after initialisation. I'll update this with a better fix.

**Comment by stedolan** [comment]
Just pushed a version that (hopefully) fixes the bug without introducing a new one.

It removes the optimised caml_callbackN for bytecode, and instead uses fixed code sequences for caml_callback1/2/3, and iterates caml_callback3 for more arguments. (This is the mechanism used by native code callbacks).

**Review by ghost — APPROVED** [review]
This is smart and looks correct to me. And it still appears to work.

**Comment by kayceesrk** [comment]
This looks correct to me. However, I'm not an expert on this part of the runtime. Perhaps a review from someone who understands this better would be useful. 

@Octachron, we should consider this for 5.3.

**Review by kayceesrk — APPROVED** [review]

**Comment by gasche** [comment]
My impression is that @xavierleroy would be the right person to look at this.

**Comment by xavierleroy** [comment]
Thanks for pinging me.  From a quick look, I agree we have a problem here, but I'm a bit sad about mirroring the native-code behavior (using "slices" of at most 3 arguments for callbacks) in the bytecode behavior.  My feeling is that it's time to change a bit the API of `caml_interprete` to make bytecode callbacks easier.  Can I have a couple of weeks to look into this before this PR is merged?

**Comment by xavierleroy** [comment]
Also: just exposing  `init_callback` and `uninit_callback` functions, and calling them when threads/domains start/stop would solve the issue with fewer changes to the existing code.

**Comment by xavierleroy** [comment]
That was easier than I thought.  See #13553 for my alternate proposal.
